src/compiler/moduleSpecifiers.ts

```diff
@@ -708,6 +712,29 @@ function getAllModulePaths(
 }
 
 function getAllModulePathsWorker(info: Info, importedFileName: string, host: ModuleSpecifierResolutionHost): readonly ModulePath[] {
+    const cache = host.getModuleResolutionCache?.();
+    const links = host.getSymlinkCache?.();
+    if (cache && links && host.readFile && !pathContainsNodeModules(info.importingSourceFileName)) {
+        Debug.type<ModuleResolutionHost>(host);
+        // Cache resolutions for all `dependencies` of the `package.json` context of the input file.
+        // This should populate all the relevant symlinks in the symlink cache, and most, if not all, of these resolutions
+        // should get (re)used.
+        const state = getTemporaryModuleResolutionState(cache.getPackageJsonInfoCache(), host, {});
+        const packageJson = getPackageScopeForPath(info.importingSourceFileName, state);
+        if (packageJson) {
+            const deps = packageJson.contents.packageJsonContent.dependencies;
+            if (deps && typeof deps === "object") {
+                const toResolve = getOwnKeys(deps as MapLike<unknown>);
+                for (const depName of toResolve) {
+                    const cached = cache.getOrCreateCacheForNonRelativeName(depName, /*mode*/ undefined, /*redirectedReference*/ undefined).get(info.canonicalSourceDirectory);

```

Comment:
Yeah, thinking about it, I probably shouldn't check the cache here at all (except maybe to check that one exists), and should just get the resolution result to feed into the symlink cache. `resolveModuleName`'ll internally do cache checking as appropriate.

Reply:
> Also i think you need to get redirectedReference from importngSourceFileName

Unreachable from here. I can get an actual redirected path form the host for the `importngSourceFileName`, if there is one, but the actual redirected reference object isn't available/exposed via the host. But also, I shouldn't really need it. A resolution of a nonrelative name from the og filepath vs the redirected path should result in the same thing, barring funky edgecases that'll make our emit assumptions invalid anyway. I could rerun the resolution from both paths... but I could also just run the resolution from the `package.json` itself. I think I like that more, honestly, and will swap to that. This is all just a heuristic to discover symlinks we haven't found yet from program construction anyway.

Reply:
The reason we have redirected references is to cache the resolutions depending on module resolution affecting options. between projects. `importngSourceFileName`  is going to be file that is going to be file that is going to be emitted (for d.ts) and from correct project (for auto import provider) so is it safe to use `options` here? If so we should pass those as otherwise we dont lookup cached result if things dont match up